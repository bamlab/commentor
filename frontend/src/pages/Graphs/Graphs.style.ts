import styled from 'styled-components';

export const StyledGraphs = styled.div`
  display: flex;
  flex-direction: column;
`;
