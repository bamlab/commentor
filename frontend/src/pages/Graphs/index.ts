export { default } from './Graphs.wrap';
