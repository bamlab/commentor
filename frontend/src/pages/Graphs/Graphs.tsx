import React, { useState, useEffect } from 'react';
import { StyledGraphs } from './Graphs.style';
import { TagType } from 'redux/Tag';
import { Doughnut } from 'react-chartjs-2';
import { CommentType } from 'redux/Comment';
import Toggle from 'components/Toggle';
import BarChart from 'components/BarChart';

interface IProps {
  tags: TagType[];
  comments: CommentType[];
  loadTags: () => void;
  loadComments: (filters: { repositoryIds: number[] }) => void;
  repositoryIds: number[];
}

const Graphs = React.memo<IProps>(props => {
  const [graphToggle, useGraphToggle] = useState('BAR_CHART');

  useEffect(() => {
    props.loadTags();
    props.loadComments({ repositoryIds: props.repositoryIds });
    // eslint-disable-next-line
  }, []);

  const getTagLabels = (tags: TagType[]): string[] => {
    let data = tags.map(data => data.code);
    data.push('⁉️'); // add no label label
    return data;
  };

  const getTagColor = (tags: TagType[]): string[] => {
    let data = tags.map(data => data.color);
    data.push('#000'); // add no label color
    return data;
  };

  // ths is best-effort 😇
  const getCommentData = (comments: CommentType[], tags: TagType[]): number[] => {
    const labels = tags.map(data => data.code);
    let data: number[] = new Array(labels.length + 1); // add no label label
    data.fill(0);
    comments.forEach(comment => {
      let noMatch = true;
      labels.forEach((label, index) => {
        if (comment.body.match(label)) {
          noMatch = false;
          data[index] += 1;
        }
      });
      if (noMatch) {
        data[labels.length] += 1;
      }
    });
    return data;
  };

  const data = {
    labels: getTagLabels(props.tags),
    datasets: [
      {
        data: getCommentData(props.comments, props.tags),
        backgroundColor: getTagColor(props.tags),
        hoverBackgroundColor: getTagColor(props.tags).map(() => '#51d'),
      },
    ],
  };
  const renderGraph = () => {
    switch (graphToggle) {
      case 'BAR_CHART':
        return <BarChart />;
      case 'DOUGHNUT_CHART':
        return <Doughnut data={data} />;
      default:
        return <div>error lol</div>;
    }
  };

  return (
    <StyledGraphs>
      <Toggle
        firstOption={'BAR_CHART'}
        secondOption={'DOUGHNUT_CHART'}
        value={graphToggle}
        onSelect={useGraphToggle}
      />
      {renderGraph()}
    </StyledGraphs>
  );
});

export default Graphs;
