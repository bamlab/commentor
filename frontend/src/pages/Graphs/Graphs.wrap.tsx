import { connect } from 'react-redux';
import { RootState } from 'redux/types';
import Graphs from './Graphs';
import { Dispatch } from 'redux';

const mapStateToProps = (state: RootState) => ({});

const mapDispatchToProps = (dispatch: Dispatch) => ({});

export default connect(
  mapStateToProps,
  mapDispatchToProps,
)(Graphs);
