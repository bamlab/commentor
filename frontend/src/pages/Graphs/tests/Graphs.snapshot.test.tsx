import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import Graphs from '../Graphs';

/*
Snapshot tests allow you to easily lock the comportment of a component.
Given props, it renders the component and compares it to the saved snapshot.
*/
describe('[Snapshot] <Tags />', () => {
  const props = {
    tags: [],
    comments: [],
    loadComments: jest.fn(),
    loadTags: jest.fn(),
  };
  it('should render Graphs page', () => {
    const wrapper = shallow(<Graphs {...props} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
