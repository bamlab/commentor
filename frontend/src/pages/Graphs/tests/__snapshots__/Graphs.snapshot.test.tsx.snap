// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Snapshot] <Tags /> should render Graphs page 1`] = `
<Graphsstyle__StyledGraphs>
  COUCOU
</Graphsstyle__StyledGraphs>
`;
